modules/misc/src/postprocessors/ThermocoupleSensorPostprocessor.C

```diff
@@ -0,0 +1,302 @@
+    }
+    Real _input_signal_delayed =
+        getDelayedInputSignal(delay_value, _time_values, _input_signal_values);
+
+    if (_method == "numerical")
+    {
+      // computing integral term
+      Real term_for_integration = _input_signal + signalToNoise_value * noise_value;
+      _integrand.push_back(term_for_integration);
+      _integration_value = getIntegral(_integrand, _time_values);
+
+      // output
+      Real proportional_value = _input_signal_delayed + signalToNoise_value * noise_value;
+      _sensor_value = drift_value +
+                      efficiency_value * (_proportional_weight * proportional_value +
+                                          _integral_weight * _integration_value) +
+                      uncertainty_value;
+    }
+
+    else if (_method == "lumped")
+    {
+      if (_t == _time_values[0])
+        _sensor_value = uncertainty_value + drift_value + 0;
+      else
+      {
+        _sensor_value = _sensor_value_old +
+                        (efficiency_value * (_input_signal + signalToNoise_value * noise_value) -
+                         _sensor_value_old) *
+                            (1 - exp(-_t / delay_value)) +
+                        uncertainty_value + drift_value;
+      }
+      _sensor_value_old = _sensor_value - uncertainty_value - drift_value;
+    }
+
+    else
+      mooseError("The entered method is invalid.");
+  }
+}
+
+PostprocessorValue
+ThermocoupleSensorPostprocessor::getValue() const
+{
+  return _sensor_value;
+}
+
+Real
+ThermocoupleSensorPostprocessor::getIntegral(std::vector<Real> integrand,
+                                             std::vector<Real> time_values)
+{
+  Real integration_value;
+  Real delay_value = getDelay();
+  std::vector<Real> exponential_term;
+  // computing vector of exponential term
+  for (size_t i = 0; i < time_values.size(); ++i)
+    exponential_term.push_back(exp(-(_t - time_values[i]) / delay_value) / delay_value);
+
+  // calculate the total integrand vector including previous integrand and exponential terms
+  integrand = elementwiseMultiply(integrand, exponential_term);
+
+  // if there are more than two datapoints, do simpson's 3/8 rule for integration
+  // else do trapezoidal rule
+  if (time_values.size() > 2)
+  {
+    SplineInterpolation spline(time_values, integrand);
+    // number of intervals
+    Real n;
+
+    // if number of datapoints is less than 30, use spline interpolation to get 30 intervals
+    if (time_values.size() < 30)
+      n = 30;
+    else
+    {
+      n = time_values.size();
+      // if interval is not a multiple of 3, make it
+      while (static_cast<int>(n) % 3 != 0)
+        n = n + 1;
+    }
+
+    // distance between time values
+    Real h = (time_values.back() - time_values[0]) / n;
+    // time vector for simpson integration
+    vector<Real> time_vec_simp;
+    // integrand vector for simpson integration
+    vector<Real> integrand_vec_simp;
+
+    for (Real i = 0; i < n + 1; i++)
```

Comment:
```suggestion
    // Add more points using spline interpolation
    for (Real i = 0; i < n + 1; i++)
```

Reply:
tbh I m not sure what's the point of adding interpolated points with a spline

seems like it would create oscillations if we got two points very close to each other with different values

Reply:
otherwise the integration accuracy seems to be bad